src/journal/journal-file.c

```diff
@@ -4034,6 +4068,26 @@ bool journal_file_rotate_suggested(JournalFile *f, usec_t max_file_usec) {
                         return true;
                 }
 
+        /* If there are too many hash collisions somebody is most likely playing games with us. Hence, if our
+         * longest chain is longer than 2/3rds of the hash table size, let's suggest rotation. */
+        if (JOURNAL_HEADER_CONTAINS(f->header, data_hash_chain_depth) &&
+            le64toh(f->header->data_hash_chain_depth) * 3 > (le64toh(f->header->data_hash_table_size) / sizeof(HashItem)) * 2) {
+                log_debug("Data hash table of %s has deepest hash chain of length %" PRIu64 " (at %" PRIu64 " hash table slots), suggesting rotation.",
+                          f->path,
+                          le64toh(f->header->data_hash_chain_depth),
+                          le64toh(f->header->data_hash_table_size) / sizeof(HashItem));
+                return true;
+        }
+
+        if (JOURNAL_HEADER_CONTAINS(f->header, field_hash_chain_depth) &&
+            le64toh(f->header->field_hash_chain_depth) * 3 > (le64toh(f->header->field_hash_table_size) / sizeof(HashItem)) * 2) {
+                log_debug("Field hash table of %s has deepest hash chain of length at %" PRIu64 " (at %" PRIu64 " hash table slots), suggesting rotation.",
+                          f->path,
+                          le64toh(f->header->field_hash_chain_depth),
+                          le64toh(f->header->field_hash_table_size) / sizeof(HashItem));
+                return true;
```

Comment:
tried, but the code would be twice as long, since it takes so many parameters.

i instead opted to drop display of the hash table size, to make the expression much shorter. after all, we don't care about the hash table size here anymore. I hope with making this shorter this is good enough now.